src/inmanta/ast/statements/__init__.py

```diff
@@ -149,15 +234,103 @@ def get_dataflow_node(self, graph: DataflowGraph) -> dataflow.NodeReference:
 
 
 class Resumer(ExpressionStatement):
+    """
+    Resume on a set of requirement variables' values when they become ready (i.e. they are complete).
+    """
+
     def resume(self, requires: Dict[object, object], resolver: Resolver, queue: QueueScheduler, target: ResultVariable) -> None:
         pass
 
 
 class RawResumer(ExpressionStatement):
+    """
+    Resume on a set of requirement variables when they become ready (i.e. they are complete).
+    """
+
     def resume(self, requires: Dict[object, ResultVariable], resolver: Resolver, queue: QueueScheduler) -> None:
         pass
 
 
+class VariableReferenceHook(RawResumer):
+    """
+    Generic helper class for adding a hook to a variable (ResultVariable) object. Supports both plain variables and instance
+    attributes. Calls variable resumer with the variable object as soon as it's available.
+    """
+
+    def __init__(
+        self,
+        instance: Optional["Reference"],
+        name: str,
+        variable_resumer: "VariableResumer",
+    ) -> None:
+        super().__init__()
+        self.instance: Optional["Reference"] = instance
+        self.name: str = name
+        self.variable_resumer: "VariableResumer" = variable_resumer
+
+    def schedule(self, resolver: Resolver, queue_scheduler: QueueScheduler) -> None:
```

Comment:
Moving scheduling out of the constructor (as is done in the ExecutionUnit side of things) is an improvement.